src/test/ui/lint/unused/unused-async.rs

```diff
@@ -1,24 +1,43 @@
 // edition:2018
-// run-pass
-#![allow(dead_code)]
+#![deny(unused_must_use)]
+
 
 #[must_use]
-//~^ WARNING `must_use`
-async fn test() -> i32 {
+async fn foo() -> i32 {
     1
 }
 
+#[must_use]
+fn bar() -> impl std::future::Future<Output=i32> {
```

Comment:
Yeah, I think we should limit this to just async fn's for now. Generalizing `#[must_use]` to look through types like `Result` and `Future` is going to require, I think, a pretty significant change (I think niko and I talked about basically needing auto-trait semantics?), and we should probably hold off on any half-measures that may make that more difficult (unless we decide its too difficult)

Reply:
*Stopping* this from working would require changing the `FnDecl` representation of all functions with information about the asyncness. Right now that is only tracked up to the HIR. I could make the lint check the `fn` item to see if it's `async` or not, but that only works for functions of the local crate. The three options here are: what it does now, and always propagate `must_use` from any `fn` to the `Future::Output` it might return, make `must_use` on `async fn` only work for local crates, or modify `ty::FnDecl` to track more metadata. I'm leaning towards the first case, despite the incorrect mental model it might imply to users for `Result` and `Option`.

Reply:
I tend to agree, but I also don't know if I should be approving decisions about where a lint applies. I started a [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/.23.5Bmust_use.5D.20on.20async.20fn.20and.20-.3E.20impl.20Future.20.23100633) to get some t-lang feedback first.